src/bio/bio_context.c

```diff
@@ -477,39 +477,46 @@ bio_blob_delete(uuid_t uuid, struct bio_xs_context *xs_ctxt)
 }
 
 static int
-bio_rw_iov(struct bio_io_context *ioctxt, bio_addr_t addr, daos_iov_t *iov,
-	   bool update)
+bio_rwv(struct bio_io_context *ioctxt, struct bio_sglist *bsgl_in,
+	daos_sg_list_t *sgl, bool update)
 {
-	struct bio_desc		*biod;
 	struct bio_sglist	*bsgl;
-	unsigned int		 iod_cnt = 1;
-	int			 rc;
+	struct bio_desc		*biod;
+	int			 i, rc;
 
 	/* allocate blob I/O descriptor */
-	biod = bio_iod_alloc(ioctxt, iod_cnt, update);
+	biod = bio_iod_alloc(ioctxt, 1 /* single bsgl */, update);
 	if (biod == NULL)
 		return -DER_NOMEM;
 
-	/* setup bio sgl in bio descriptor */
-	bsgl = bio_iod_sgl(biod, 0); /* bsgl = &biod->bd_sgls[0] */
-	rc = bio_sgl_init(bsgl, iod_cnt); /* sets up bsgl->bs_iovs */
+	/*
+	 * copy the passed in @bsgl_in to the bsgl attached on bio_desc,
+	 * since we don't want following bio ops change caller's bsgl.
+	 */
+	bsgl = bio_iod_sgl(biod, 0);
+
+	rc = bio_sgl_init(bsgl, bsgl_in->bs_nr);
 	if (rc)
-		goto out; /* rc = -DER_NOMEM */
+		goto out;
 
-	/* store byte offset and device type */
-	bsgl->bs_iovs[0].bi_addr = addr;
-	bsgl->bs_iovs[0].bi_data_len = iov->iov_len;
-	bsgl->bs_nr_out++;
+	for (i = 0; i < bsgl->bs_nr; i++) {
+		D_ASSERT(bsgl_in->bs_iovs[i].bi_buf == NULL);
+		D_ASSERT(bsgl_in->bs_iovs[i].bi_data_len != 0);
+		bsgl->bs_iovs[i] = bsgl_in->bs_iovs[i];
+	}
+	bsgl->bs_nr_out = bsgl->bs_nr;
 
 	/* map the biov to DMA safe buffer, fill DMA buffer if read operation */
 	rc = bio_iod_prep(biod);
 	if (rc)
 		goto out;
-	D_ASSERT(bsgl->bs_iovs[0].bi_buf != NULL);
 
-	/* copy data from/to iov and DMA safe buffer for write/read */
-	bio_memcpy(biod, addr.ba_type, bsgl->bs_iovs[0].bi_buf,
-		   iov->iov_buf, iov->iov_len);
+	for (i = 0; i < bsgl->bs_nr; i++)
+		D_ASSERT(bsgl->bs_iovs[i].bi_buf != NULL);
```

Comment:
it is a little too expensive to call assertion in a loop, we could move the assertion to bio_idod_copy()->copy_one?